zowe/zlux-server-framework
# [WIP] Sync two NodeJS servers, failover experiment

### New environment variables shared across zlux instances:  
* ZOWE_INSTANCE, used in session cookie name
* ZOWE_SESSION_SECRET - temporary solution

### How to test a failover
0. Set global environment variables
```
ZOWE_INSTANCE="my_zowe"
ZOWE_SESSION_SECRET="my-secret"
ZLUX_MAX_WORKERS=1
```
1. Setup two zlux instances with Mediation layer enabled
2. Run API Mediation Layer
3. Run first instance - it will be Master instance, wait until it finishes startup
4. Open Eureka UI at https://localhost:10011 (user/user) and check that you see a single ZLUX instance `UP` ![image](https://user-images.githubusercontent.com/34669449/79449180-340d6280-7ffc-11ea-8679-52aa5f8bf7d2.png)
5. Run second instance - it will be a backup
6. Open Eureka UI at https://localhost:10011 and check that you see two ZLUX instances one - `UP`, and one - `OUT_OF_SERVICE`
![image](https://user-images.githubusercontent.com/34669449/79449428-a1b98e80-7ffc-11ea-880c-a9b171d5af0e.png)
7. Open Zlux UI at https://localhost:10010/ui/v1/zlux/ZLUX/plugins/org.zowe.zlux.bootstrap/web/, run an app, etc.
8.  Stop first instance(Backup instance will detect that Master has failed, and will take into service)
9. Open Eureka UI at https://localhost:10011 and check that you see the second ZLUX instance UP, and the first - `OUT_OF_SERVICE` 
![image](https://user-images.githubusercontent.com/34669449/79449951-9450d400-7ffd-11ea-83cc-2eea11272796.png) the first will disappear after a timeout
![image](https://user-images.githubusercontent.com/34669449/79450001-ac285800-7ffd-11ea-809f-8312e7d19cfe.png)
7. Ensure that Zlux UI still works
